main.py

```diff
@@ -0,0 +1,347 @@
+# Copyright 2017 Google, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This main python file contains the backend logic of the website, including
+the reception of Cloud Pub/Sub messages and communication with GCS
+and Datastore.
+"""
+
+import collections
+import json
+import logging
+import os
+import urllib
+
+import cloudstorage
+import jinja2
+import webapp2
+from google.appengine.api import images
+from google.appengine.ext import blobstore
+from google.appengine.ext import ndb
+import googleapiclient.discovery
+
+# Constants. Note: Change THUMBNAIL_BUCKET and PHOTO_BUCKET to
+# be applicable to your project.
+THUMBNAIL_BUCKET = 'photo-thumbnails-bucket'
+PHOTO_BUCKET = 'photo-album-bucket'
+NUM_NOTIFICATIONS_TO_DISPLAY = 50
+MAX_LABELS = 5
+
+# Set up jinja2 for HTML templating.
+template_dir = os.path.join(os.path.dirname(__file__), 'templates')
+jinja_environment = jinja2.Environment(
+    loader=jinja2.FileSystemLoader(template_dir))
+
+
+class Notification(ndb.Model):
+    '''Describes a Notification to be displayed on the home/news
+    feed page. String message, date of posting, and generation number
+    to prevent the display of repeated notifications.'''
+    message = ndb.StringProperty()
+    date = ndb.DateTimeProperty(auto_now_add=True)
+    generation = ndb.StringProperty()
+
+
+class ThumbnailReference(ndb.Model):
+    '''Describes a ThumbnailReference, which holds information about a given
+    thumbnail (not the thumbnail itself).
+    thumbnail_name: same as photo name.
+    thumbnail_key: used to distinguish similarly named photos. Includes
+    name and generation number.
+    labels: list of label_names that apply to the photo.
+    original_photo: url of the original photo, stored in GCS.'''
+    thumbnail_name = ndb.StringProperty()
+    thumbnail_key = ndb.StringProperty()
+    date = ndb.DateTimeProperty(auto_now_add=True)
+    labels = ndb.StringProperty(repeated=True)
+    original_photo = ndb.StringProperty()
+
+
+class MainHandler(webapp2.RequestHandler):
+    '''Home/news feed page (notification listing).'''
+    def get(self):
+        # Fetch all notifications in reverse date order.
+        notifications = Notification.query().order(
+            -Notification.date).fetch(NUM_NOTIFICATIONS_TO_DISPLAY)
+        template_values = {'notifications': notifications}
+        template = jinja_environment.get_template("notifications.html")
```

Comment:
style nit: use single quotes throughout for strings (except docstrings which use `"""`).